### PR TITLE
RL doc string

### PR DESCRIPTION
**Pull request by lihuoran**
# Description

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.6
  - [ ] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/training/trainer.py [hunk 1]
```diff
@@ -60,12 +88,20 @@ def register_agent2policy(self, agent2policy: Dict[Any, str]) -> None:
...
+
+        Args:
+            global_policy_creator (Dict[str, Callable[[str], RLPolicy]]): Dict that contains the creators for all
+                policies.
+        """
         raise NotImplementedError
 
     @abstractmethod
     def build(self) -> None:
+        """Build the trainer. This has to be called before the trainer can actually run.
```
Comment (ysqyang) [inline maro/rl/training/trainer.py:103]: Create the required train-ops and replay memory. This should be called before invoking ``train`` or ``train_as_task``.

File: maro/rl/training/trainer.py [hunk 2]
```diff
@@ -75,22 +111,47 @@ def train(self) -> None:
         raise NotImplementedError
 
     async def train_as_task(self) -> None:
-        """Run a training step to update all the policies that this trainer is responsible for.
+        """Run a training step, in the async fashion, to update all the policies that this trainer is responsible for.
```
Comment (ysqyang) [inline maro/rl/training/trainer.py:114]: Update all policies managed by the trainer as an asynchronous task.

File: maro/rl/training/trainer.py [hunk 3]
```diff
@@ -75,22 +111,47 @@ def train(self) -> None:
...
+            exp_element (ExpElement): Experiences.
+        """
         raise NotImplementedError
 
     def set_proxy_address(self, proxy_address: Tuple[str, int]) -> None:
         self._proxy_address = proxy_address
 
     @abstractmethod
     def get_local_ops_by_name(self, name: str) -> AbsTrainOps:
+        """Generate a train ops according to the given ops name.
```
Comment (ysqyang) [inline maro/rl/training/trainer.py:134]: Create an ``AbsTrainOps`` instance with a given name.

File: maro/rl/training/algorithms/ac.py [hunk 4]
```diff
@@ -93,18 +105,42 @@ def _get_critic_loss(self, batch: TransitionBatch) -> torch.Tensor:
...
+        """Update the critic according to the given gradients.
+
+        Args:
+            grad_dict (dict): Gradients.
         """
         self._v_critic_net.train()
         self._v_critic_net.apply_gradients(grad_dict)
 
     def _get_actor_loss(self, batch: TransitionBatch) -> torch.Tensor:
+        """Get the actor loss of the given batch.
```
Comment (ysqyang) [inline maro/rl/training/algorithms/ac.py:136]: Compute the actor loss for a batch.

File: maro/rl/training/algorithms/ac.py [hunk 5]
```diff
@@ -132,28 +168,52 @@ def _get_actor_loss(self, batch: TransitionBatch) -> torch.Tensor:
 
     @remote
     def get_actor_grad(self, batch: TransitionBatch) -> Dict[str, torch.Tensor]:
+        """Get the actor gradients of the given batch.
```
Comment (ysqyang) [inline maro/rl/training/algorithms/ac.py:171]: Compute the actor network's gradient for a batch.

File: maro/rl/training/algorithms/ac.py [hunk 6]
```diff
@@ -132,28 +168,52 @@ def _get_actor_loss(self, batch: TransitionBatch) -> torch.Tensor:
...
-            "critic": self._v_critic_net.get_state()
+            "critic": self._v_critic_net.get_state(),
         }
 
     def set_state(self, ops_state_dict: dict) -> None:
         self._policy.set_state(ops_state_dict["policy"])
         self._v_critic_net.set_state(ops_state_dict["critic"])
 
     def _preprocess_batch(self, batch: TransitionBatch) -> TransitionBatch:
+        """Preprocess the batch to get the return & advantages.
```
Comment (ysqyang) [inline maro/rl/training/algorithms/ac.py:209]: to get returns & advantages

File: maro/rl/training/algorithms/ac.py [hunk 7]
```diff
@@ -85,6 +89,14 @@ def __init__(
         self._v_critic_net.to(self._device)
 
     def _get_critic_loss(self, batch: TransitionBatch) -> torch.Tensor:
+        """Get the critic loss of the given batch.
```
Comment (ysqyang) [inline maro/rl/training/algorithms/ac.py:92]: Compute the critic loss for a given batch.

File: maro/rl/rollout/env_sampler.py [hunk 8]
```diff
@@ -236,29 +240,27 @@ def __init__(
...
+            Global state (np.ndarray)
+            Dict of agent states (Dict[Any, np.ndarray])
         """
         raise NotImplementedError
 
     @abstractmethod
     def _translate_to_env_action(self, action_dict: Dict[Any, np.ndarray], event: object) -> Dict[Any, object]:
-        """
-        Translation the actions into the format that the env could recognize.
+        """Translation the actions into the format that the env could recognize.
```
Comment (ysqyang) [inline maro/rl/rollout/env_sampler.py:259]: Translate model-generated actions into an object that can be executed by the env.

File: maro/rl/rollout/env_sampler.py [hunk 9]
```diff
@@ -236,29 +240,27 @@ def __init__(
 
     @abstractmethod
     def _get_global_and_agent_state(
-        self, event: object, tick: int = None
+        self, event: object, tick: int = None,
     ) -> Tuple[Optional[np.ndarray], Dict[Any, np.ndarray]]:
-        """
-        Get global state and dict of agent states.
+        """Get global state and dict of agent states.
```
Comment (ysqyang) [inline maro/rl/rollout/env_sampler.py:245]: Get the global and individual agents' states.

File: maro/rl/rollout/env_sampler.py [hunk 10]
```diff
@@ -17,44 +17,40 @@
...
 
     def choose_actions(self, state_by_agent: Dict[Any, np.ndarray]) -> Dict[Any, np.ndarray]:
-        """
-        Choose action according to the given (observable) states of all agents.
+        """Choose action according to the given (observable) states of all agents.
 
         Args:
-            state_by_agent (Dict[Any, np.ndarray]): Agent state dict for all agents. Use agent name as the key to fetch
-                states from this dict.
+            state_by_agent (Dict[Any, np.ndarray]): Agent state dict for all agents.
```
Comment (ysqyang) [inline maro/rl/rollout/env_sampler.py:49]: Dictionary containing each agent's state vector. The keys are agent names.

File: maro/rl/model/policy_net.py [hunk 11]
```diff
@@ -113,21 +115,20 @@ def get_action_probs(self, states: torch.Tensor) -> torch.Tensor:
...
-            Lop-probability matrix with shape [batch_size, action_num]
+            logps (torch.Tensor): Lop-probability matrix with shape [batch_size, action_num].
         """
         return torch.log(self.get_action_probs(states))
 
     @abstractmethod
     def _get_action_probs_impl(self, states: torch.Tensor) -> torch.Tensor:
-        """
-        Implementation of `get_action_probs`.
+        """Implementation of `get_action_probs`. This method contains core logic of the net, and should be determined
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/model/policy_net.py:130]: Changed to "The core logic of a discrete policy net should be implemented here.". I personally prefer "should". "is" sounds like the logic has already been implemented.

File: maro/rl/training/algorithms/maddpg.py [hunk 12]
```diff
@@ -119,34 +155,62 @@ def _get_critic_loss(self, batch: MultiTransitionBatch, next_actions: List[torch
...
         self._q_critic_net.apply_gradients(grad_dict)
 
     def _get_actor_loss(self, batch: MultiTransitionBatch) -> torch.Tensor:
+        """Get the actor loss of the given batch.
+
+        Args:
+            batch (MultiTransitionBatch): Batch.
+
+        Returns:
+            loss (torch.Tensor): The actor loss of the batch.
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/training/algorithms/maddpg.py:212]: I prefer "of" here. For example, this [link](https://towardsdatascience.com/batch-mini-batch-stochastic-gradient-descent-7a62ecba642a) mentions "gradients of the batch". What is your opinion?

File: maro/rl/model/policy_net.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ysqyang) [reply maro/rl/model/policy_net.py:130]: Sounds good

File: maro/rl/training/algorithms/maddpg.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ysqyang) [reply maro/rl/training/algorithms/maddpg.py:212]: Yes, I guess "of" is fine.

File: maro/rl/policy/abs_policy.py [hunk 13]
```diff
@@ -82,24 +77,22 @@ def _rule(self, states: object) -> object:
...
 class RLPolicy(AbsPolicy, metaclass=ABCMeta):
-    """
-    Reinforcement learning policy.
+    """Reinforcement learning policy.
+
+    Args:
+        name (str): Name of the policy.
+        state_dim (int): Dimension of states.
+        action_dim (int): Dimension of actions.
+        trainable (bool, default=True): Whether this policy is trainable.
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/policy/abs_policy.py:86]: Yes, and it default value is True. Users do not to be aware of this setting in most cases.

File: maro/rl/policy/abs_policy.py [hunk 14]
```diff
@@ -54,20 +51,18 @@ def set_name(self, name: str) -> None:
...
 
-    def get_actions(self, states: object) -> object:
+    def get_actions(self, states: object) -> None:
         return None
 
 
 class RuleBasedPolicy(AbsPolicy, metaclass=ABCMeta):
-    """
-    Rule-based policy. The user should implement the rule of this policy, and a rule-based policy is not trainable.
+    """Rule-based policy. The user should define the rule of this policy, and a rule-based policy is not trainable.
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/policy/abs_policy.py:65]: Statistics-based policy should not belong to rule-based policies. If it is necessary, we should & need to design a new abstraction for it.
